transmart-copy/src/main/groovy/org/transmartproject/copy/table/Patients.groovy

```diff
@@ -149,4 +149,38 @@ class Patients {
         log.info "${insertCount} patients inserted."
     }
 
+    void removeObservations(String rootPath, Set<Long> trialVisitNums) {
+        LinkedHashMap<String, Class> header
+        def patientsFile = new File(rootPath, PATIENT_MAPPING_TABLE.fileName)
+        patientsFile.withReader { reader ->
+            def tsvReader = Util.tsvReader(reader)
+            tsvReader.eachWithIndex { String[] data, int i ->
+                if (i == 0) {
+                    header = Util.verifyHeader(PATIENT_MAPPING_TABLE.fileName, data, patientMappingColumns)
+                    return
+                }
+                def patientData = Util.asMap(header, data)
+                def patientIde = patientData['patient_ide'] as String
+                def patientIdeSource = patientData['patient_ide_source'] as String
+                def key = "${patientIdeSource}:${patientIde}".toString()
+                def patientNum = subjectIdToPatientNum[key]
+
+                removeObservationsForPatientAndTrials(patientNum, trialVisitNums)
+            }
+        }
+    }
+
+    private void removeObservationsForPatientAndTrials(Long patientNum, Set<Long> trialVisitNums) {
+        if (!trialVisitNums) {
+            return
+        }
+        int observationCount = database.namedParameterJdbcTemplate.update(
+                """delete from ${Observations.TABLE} where
```

Comment:
This could be a single delete for all these patients, right? (but then in chunks of max_parameters)

Reply:
In the `void load(String rootPath)` function, the existing patients are already available.